### PR TITLE
[API300] Add support to Scopes API300 new feature

### LINKED ISSUE #133: [API300] Add support to Scopes
### Scenario/Intent
Add support to Scopes resource on API300.

The `API300::[Synergy/C7000]::ServerHardware#patch` method was made **only** to operate with **Scopes** but it's currently not being supported by the Ruby SDK.

### Environment Details
 - **OneView SDK Version:** 3.1.0
 - **OneView Client API Version:** API300

### Steps to Reproduce
Try to use the API300 Server Hardware patch as described in the REST API documentation.

### Expected Result
I should be able to use the Ruby SDK to gather the Scope information needed.

### Actual Result
I actually can't do this since it does not exists yet.

### PR DESCRIPTION
**Pull request by ricardogpsf**
### Description
Support to Scopes API300 new feature for C7000 and Synergy

### Issues Resolved
Fixes #133 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass (`$ rake test`).
- [ ] New functionality has been documented in the README if applicable.
  - [x] New functionality has been thoroughly documented in the examples (please include helpful comments).
- [x] Changes are documented in the CHANGELOG.

### CODE REVIEW

**Comment by ricardogpsf** [comment]
@tmiotto @fgbulsoni @jsmartt Could you review, please?! =)

File: lib/oneview-sdk/resource/api300/c7000/server_hardware.rb [hunk 1]
```diff
@@ -75,6 +76,35 @@ def patch(operation, path, value = nil)
...
+          scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
+          patch('remove', "/scopeUris/#{scope_index}", nil) if scope_index
+        end
+
+        # Change the list of scopes in the enclosure
+        # @param [Array[OneviewSDK::API300::C7000::Scope]] scopes The scopes list (or scopes separeted by comma)
+        # @raise [OneviewSDK::IncompleteResource] if the uri of each scope is not set
+        def replace_scopes(*scopes)
+          scopes = scopes.flatten
+          uris = scopes.map do |scope|
```
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/server_hardware.rb]: This scopes flatten and map is repeated all over the code. Please create a helper method to it.

File: spec/integration/resource/api300/synergy/server_hardware/update_spec.rb [hunk 2]
```diff
@@ -4,37 +4,78 @@
...
+
+    it 'should add scope' do
+      expect { item.add_scope(scope_1) }.to_not raise_error
+      item.refresh
+      expect(item['scopeUris']).to match_array([scope_1['uri']])
+    end
+  end
+
+  describe '#replace_scopes' do
+    context 'when scope has not uri' do
```
Comment (tmiotto) [inline spec/integration/resource/api300/synergy/server_hardware/update_spec.rb]: Did you mean *"has no URI"* . 

Please check it, it's spread all over the code.

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb [hunk 3]
```diff
@@ -25,22 +25,37 @@ def initialize(client, params = {}, api_ver = nil)
...
+          scope.ensure_uri
+          scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
+          patch('remove', "/scopeUris/#{scope_index}", nil) if scope_index
+        end
+
+        # Change the list of scopes in the enclosure
+        # @param [Array[OneviewSDK::API300::C7000::Scope]] scopes The scopes list (or scopes separeted by comma)
+        # @raise [OneviewSDK::IncompleteResource] if the uri of each scope is not set
+        def replace_scopes(*scopes)
+          scopes = scopes.flatten
```
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: Because when we use flatten! the scopes object (parameter) is changed, and maybe not be expected by who called the api.

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb [hunk 4]
```diff
@@ -25,22 +25,37 @@ def initialize(client, params = {}, api_ver = nil)
...
+        end
+
+        # Remove one scope from the enclosure
+        # @param [OneviewSDK::API300::C7000::Scope] scope The scope resource
+        # @raise [OneviewSDK::IncompleteResource] if the uri of scope is not set
+        def remove_scope(scope)
+          scope.ensure_uri
+          scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
+          patch('remove', "/scopeUris/#{scope_index}", nil) if scope_index
+        end
```
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: Good. We could return NotFound error. What do you think about it?

Comment (jsmartt) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: I'm not sure. Maybe it's OK to return nil (or false), since it technically is removed (maybe it was never there). I suppose it's fine if you think it's good.

File: lib/oneview-sdk/resource/api300/c7000/scope.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        # Set data and save to OneView
+        # @param [Hash] attributes The attributes to add/change for this resource (key-value pairs)
+        # @raise [OneviewSDK::IncompleteResource] if the client or uri is not set
+        # @raise [StandardError] if the resource save fails
+        # @return [Resource] self
+        def update(attributes = {})
+          set_all(attributes)
+          ensure_client && ensure_uri
+          options = {
+            'If-Match' => @data['eTag'],
```
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/scope.rb]: In the documentation of scopes, for update and delete methods, the `If-Match ` tag for header is described as required http://apidocs.vse.rdlabs.hpecorp.net/fusion/300/c7000/public/#rest/scopes

When I called the method of rest-api without `If-Match` tag in header, I received a error about it.

File: lib/oneview-sdk/resource/api300/c7000/scope.rb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        # Adds resource assignments
+        # @param [Array] resources The array of resources (or resources separated by comma)
+        # @raise [OneviewSDK::IncompleteResource] if the client or uri is not set
+        def set_resources(*resources)
+          change_resources_assignments(add_resources: resources.flatten)
+        end
+
+        # Removes resource assignments
+        # @param [Array|Strings] resources The array of resources (or resources separated by comma)
```
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/scope.rb]: I'm working in that now. It will be `@param [Array]`. =)
But client can use any number of strings (without use of the array)

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb (hunk 4 above)
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: I liked the idea, maybe we could to return true or false. True when the resource was removed and false when was not removed (but did not occur error)

File: lib/oneview-sdk/resource/api300/c7000/scope.rb (hunk 5 above)
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/scope.rb]: why?

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb (hunk 3 above)
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: Well, the splat creates a copy of what you're passing as a parameter (the arguments array) for some reasons, first because the `scopes` is not actually the parameters you're passing, but the arguments array created by ruby, second if you pass actually an array it'll treat as one parameter and you'll be actually flattening the arguments array, and at last if you splat the input array to the method (i.e. `replace_scopes(*my_input)` ) the code still needs the argument array, so it is still a copy of it.

TL:DR It won't change anything since you're using the splat operator.

It just seems weird to me having to assign it to the same variable when using splat, but not a required change.

File: lib/oneview-sdk/resource/api300/c7000/scope.rb (hunk 5 above)
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/scope.rb]: Some resources are already doing it, and some of them complain the eTag cannot be modified, and if they are persent on the body cause the request to fail. Obviously it's not the case of Scopes, but it would avoid future breakage if this change is made to it, and it would help to keep a standard over the code.

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb (hunk 3 above)
Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: Truth. The splat avoid the problem with `flatten!` overriding the original array. =)

**Comment by ricardogpsf** [comment]
The changes were done!

**Comment by jsmartt** [comment]
Looks good to me

**Comment by ricardogpsf** [comment]
@tmiotto are there still request changes that you did?

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb [hunk 7]
```diff
@@ -39,22 +39,25 @@ def add_scope(scope)
 
         # Remove one scope from the enclosure
         # @param [OneviewSDK::API300::C7000::Scope] scope The scope resource
+        # @return [Boolean] True if the scope was deleted and false if enclosure has not the scope
         # @raise [OneviewSDK::IncompleteResource] if the uri of scope is not set
         def remove_scope(scope)
           scope.ensure_uri
           scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
-          patch('remove', "/scopeUris/#{scope_index}", nil) if scope_index
+          if scope_index
```
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: It would be better to use a guard clause here:

```Ruby
return false unless scope_index
patch('remove', "/scopeUris/#{scope_index}", nil)
true
```

File: lib/oneview-sdk/resource.rb [hunk 8]
```diff
@@ -323,6 +323,15 @@ def ensure_uri
       true
     end
 
+    # Fail unless all uri of resources is set.
```
Comment (tmiotto) [inline lib/oneview-sdk/resource.rb]: I recommend to improve this comment:
1. First it would be better if it was a description of what it really does. (**e.g.** *Gets all the URIs for the specified resources*)
2. Add the YARD tags for the @param, @return and @raise

File: lib/oneview-sdk/resource/api300/c7000/server_hardware.rb [hunk 9]
```diff
@@ -86,22 +86,25 @@ def add_scope(scope)
 
         # Remove one scope from the enclosure
         # @param [OneviewSDK::API300::C7000::Scope] scope The scope resource
+        # @return [Boolean] True if the scope was deleted and false if enclosure has not the scope
         # @raise [OneviewSDK::IncompleteResource] if the uri of scope is not set
         def remove_scope(scope)
           scope.ensure_uri
           scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
-          patch('remove', "/scopeUris/#{scope_index}", nil) if scope_index
+          if scope_index
```
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/server_hardware.rb]: Same guard clause suggestion here

File: lib/oneview-sdk/resource/api300/c7000/enclosure.rb [hunk 10]
```diff
@@ -25,22 +25,40 @@ def initialize(client, params = {}, api_ver = nil)
...
+        end
+
+        # Remove one scope from the enclosure
+        # @param [OneviewSDK::API300::C7000::Scope] scope The scope resource
+        # @return [Boolean] True if the scope was deleted and false if enclosure has not the scope
+        # @raise [OneviewSDK::IncompleteResource] if the uri of scope is not set
+        def remove_scope(scope)
+          scope.ensure_uri
+          scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
+          if scope_index
```
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/enclosure.rb]: > Somehow my comment was marked in a outdated change... anyway, here it is again:

It would be better to use a guard clause here:

```Ruby
return false unless scope_index
patch('remove', "/scopeUris/#{scope_index}", nil)
true
```

File: lib/oneview-sdk/resource/api300/c7000/server_hardware.rb [hunk 11]
```diff
@@ -75,6 +76,38 @@ def patch(operation, path, value = nil)
...
+        end
+
+        # Remove one scope from the enclosure
+        # @param [OneviewSDK::API300::C7000::Scope] scope The scope resource
+        # @return [Boolean] True if the scope was deleted and false if enclosure has not the scope
+        # @raise [OneviewSDK::IncompleteResource] if the uri of scope is not set
+        def remove_scope(scope)
+          scope.ensure_uri
+          scope_index = @data['scopeUris'].find_index { |uri| uri == scope['uri'] }
+          if scope_index
```
Comment (tmiotto) [inline lib/oneview-sdk/resource/api300/c7000/server_hardware.rb]: > Somehow my comment was marked in a outdated change... anyway, here it is again:

It would be better to use a guard clause here:

```Ruby
return false unless scope_index
patch('remove', "/scopeUris/#{scope_index}", nil)
true
```

Comment (ricardogpsf) [inline lib/oneview-sdk/resource/api300/c7000/server_hardware.rb]: Done

**Comment by ricardogpsf** [comment]
the changes were done, @tmiotto
